packages/pigeon/lib/kotlin/templates.dart

```diff
@@ -0,0 +1,344 @@
+    require(!weakInstances.containsKey(identifier)) {
+      "Identifier has already been added: \$identifier"
+    }
+    val weakReference = java.lang.ref.WeakReference(instance, referenceQueue)
+    identifiers[instance] = identifier
+    weakInstances[identifier] = weakReference
+    weakReferencesToIdentifiers[weakReference] = identifier
+    strongInstances[identifier] = instance
+  }
+
+  private fun logWarningIfFinalizationListenerHasStopped() {
+    if (hasFinalizationListenerStopped()) {
+      Log.w(
+        tag,
+        "The manager was used after calls to the $_finalizationListenerClassName has been stopped."
+      )
+    }
+  }
+}
+''';
+
+/// Creates the `InstanceManagerApi` with the passed string values.
+String instanceManagerApiTemplate({
+  required String dartPackageName,
+  required String errorClassName,
+}) {
+  final String removeStrongReferenceName = makeChannelNameWithStrings(
+    apiName: _instanceManagerApiName,
+    methodName: 'removeStrongReference',
+    dartPackageName: dartPackageName,
+  );
+  final String clearName = makeChannelNameWithStrings(
+    apiName: _instanceManagerApiName,
+    methodName: 'clear',
+    dartPackageName: dartPackageName,
+  );
+  return '''
+/**
+* Generated API for managing the Dart and native `$instanceManagerClassName`s.
+*/
+private class $_instanceManagerApiName(val binaryMessenger: BinaryMessenger) {
+  companion object {
+    /** The codec used by $_instanceManagerApiName. */
+    private val codec: MessageCodec<Any?> by lazy {
+      StandardMessageCodec()
+    }
+
+    /**
+    * Sets up an instance of `$_instanceManagerApiName` to handle messages from the
+    * `binaryMessenger`.
+    */
+    fun setUpMessageHandlers(binaryMessenger: BinaryMessenger, instanceManager: $instanceManagerClassName?) {
+      run {
+        val channel = BasicMessageChannel<Any?>(binaryMessenger, "$removeStrongReferenceName", codec)
+        if (instanceManager != null) {
+          channel.setMessageHandler { message, reply ->
+            val identifier = message as Number
+            val wrapped: List<Any?> = try {
+              instanceManager.remove<Any?>(identifier.toLong())
+              listOf<Any?>(null)
+            } catch (exception: Throwable) {
+              wrapError(exception)
+            }
+            reply.reply(wrapped)
+          }
+        } else {
+          channel.setMessageHandler(null)
+        }  
+      }
+      run {
+        val channel = BasicMessageChannel<Any?>(binaryMessenger, "$clearName", codec)
+        if (instanceManager != null) {
+          channel.setMessageHandler { _, reply ->
+            val wrapped: List<Any?> = try {
+              instanceManager.clear()
+              listOf<Any?>(null)
+            } catch (exception: Throwable) {
+              wrapError(exception)
+            }
+            reply.reply(wrapped)
+          }
+        } else {
+          channel.setMessageHandler(null)
+        }
+      }
+    }
+  }
```

Comment:
a lot of this code is essentially duplicated with the host api. I'd prefer if that code could be gened from the same code. to avoid future work duplication

Reply:
This will also require changes to the Dart code, so I created https://github.com/flutter/packages/pull/6831